### PR TITLE
Chapter7

### PR DESCRIPTION
**Pull request by NurAnisAqilah**

### CODE REVIEW

**Comment by NurAnisAqilah** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 1]
```diff
@@ -1,36 +1,178 @@
...
+  letter-spacing: -1px;
+  margin-bottom: 30px;
+  text-align: center;
+  font-weight: normal;
+  color: $gray-light;
+
+}
+  p {
+  font-size: 1.1em;
+  line-height: 1.7em;
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion

p {
  font-size: 1.1em;
  line-height: 1.7em;
```

File: app/assets/stylesheets/custom.scss [hunk 2]
```diff
@@ -1,36 +1,178 @@
...
+  color: $blue;
+  text-transform: uppercase;
+  letter-spacing: -1px;
+  padding-top: 9px;
+  font-weight: bold;
+  &:hover {
+    color: white;
+    text-decoration: none;
+    }
+}
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
  font-weight: bold;
  
  &:hover {
    color: white;
    text-decoration: none;
    }
}
```

File: app/assets/stylesheets/custom.scss [hunk 3]
```diff
@@ -1,36 +1,178 @@
...
+footer {
+  margin-top: 400px;
+  padding-top: 5px;
+  border-top: 1px solid $red;
+  color: $gray-light;
+  a {
+    color: $gray;
+    &:hover{
+      color: $gray-darker;
+    }
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
  color: $gray-light;
  
  a {
    color: $gray;
    
    &:hover{
      color: $gray-darker;
    }
```

File: app/assets/stylesheets/custom.scss [hunk 4]
```diff
@@ -1,36 +1,178 @@
...
+  }
+  small {
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
  }
  
  small {
```

File: app/assets/stylesheets/custom.scss [hunk 5]
```diff
@@ -1,36 +1,178 @@
...
-    color: #fff;
-    text-transform: uppercase;
-    letter-spacing: -1px;
-    padding-top: 9px;
-    font-weight: bold;
+  }
+  ul {
+    float: right;
+    list-style: none;
+    li {
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
  }
  
  ul {
    float: right;
    list-style: none;
    
    li {
```

File: app/assets/stylesheets/custom.scss [hunk 6]
```diff
@@ -1,36 +1,178 @@
...
+  @include box_sizing;
+}
+
+input {
+  height: auto !important;
+}
+
+#error_explanation {
+  color:$red;
+  ul{
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
  color:$red;
  
  ul{
```

File: app/assets/stylesheets/custom.scss [hunk 7]
```diff
@@ -1,36 +1,178 @@
...
+    color:$red;
+    margin: 0 0 30px 0;
+  }
+}
+
+.field_with_errors {
+  @extend .has-error;
+  .form-control {
+    color: $state-danger-text;
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]:
```suggestion
  @extend .has-error;
  
  .form-control {
    color: $state-danger-text;
```

File: app/assets/stylesheets/users.scss [hunk 8]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Users controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (chungpv-1008) [inline app/assets/stylesheets/users.scss]: remove this file

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -0,0 +1,34 @@
+class UsersController < ApplicationController
+  before_action :load_user, only: [:show, :edit]
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
  before_action :load_user, only: %i(show edit)
```

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    @user = User.new
+  end
+
+  def create
+    @user = User.new user_params
+    if @user.save
+      flash[:success] = "Welcome to the Sample App"
+      redirect_to @user
+    else
+      flash[:danger] = "Please try again"
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
      flash.now[:danger] = "Please try again"
```

File: app/helpers/application_helper.rb [hunk 11]
```diff
@@ -1,2 +1,7 @@
-module ApplicationHelper
+module UsersHelper
+  def gravatar_for(user)
+    gravatar_id = Digest::MD5::hexdigest(user.email.downcase)
+    gravatar_url = "https://secure.gravatar.com/avatar/#{gravatar_id}"
+    image_tag(gravatar_url, alt: user.name, class: "gravatar")
```
Comment (chungpv-1008) [inline app/helpers/application_helper.rb]:
```suggestion
  def gravatar_for user
    gravatar_id = Digest::MD5::hexdigest user.email.downcase
    gravatar_url = "https://secure.gravatar.com/avatar/#{gravatar_id}"
    image_tag gravatar_url, alt: user.name, class: "gravatar"
```

File: app/views/layouts/application.html.erb [hunk 12]
```diff
@@ -1,18 +1,30 @@
...
+    </div>
+  </body>
   <%= render "layouts/header" %>
   <div class="container">
+    <% flash.each do |message_type, message| %>
+      <div class="alert alert-<%= message_type %>"><%= message %></div>
+    <% end %>
     <%= yield %>
+    <%= render "layouts/footer" %>
+    <%= debug(params) if Rails.env.development? %>
```
Comment (chungpv-1008) [inline app/views/layouts/application.html.erb]:
```suggestion
    <%= debug params if Rails.env.development? %>
```

File: app/views/users/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,28 @@
+<% provide :title, "Sign up" %>
+<h1>Sign up</h1>
+<p>This will be a signup page for new users.</p>
+<%= provide :title, "Sign Up" %>
```
Comment (chungpv-1008) [inline app/views/users/new.html.erb]:
```suggestion
```

File: config/routes.rb [hunk 14]
```diff
@@ -1,7 +1,11 @@
 Rails.application.routes.draw do
-  scope "(:locale)", locale: /en/ do
-    root "static_pages#home"
-    get "static_pages/help"
-    get "static_pages/contact"
-  end
+  get 'users/show'
+  get 'users/new'
```
Comment (chungpv-1008) [inline config/routes.rb]:
```suggestion
```

File: config/routes.rb [hunk 15]
```diff
@@ -1,7 +1,11 @@
...
+  root "static_pages#home"
+  get "/help", to: "static_pages#help"
+  get "/about", to: "static_pages#about"
+  get "/contact", to: "static_pages#contact"
+  get "/signup", to: "users#new"
+
+  resources :users
```
Comment (chungpv-1008) [inline config/routes.rb]:
```suggestion
  resources :users, only: %i(show new)
```

**Comment by NurAnisAqilah** [comment]
ready

File: app/assets/stylesheets/custom.scss (hunk 1 above)
Comment (chungpv-1008) [reply app/assets/stylesheets/custom.scss]: Please fix all comment before 'ready'

File: app/views/users/new.html.erb (hunk 13 above)
Comment (NurAnisAqilah) [reply app/views/users/new.html.erb]: how should I resolve this?

**Comment by NurAnisAqilah** [comment]
ready

File: app/helpers/application_helper.rb [hunk 16]
```diff
@@ -1,2 +1,7 @@
-module ApplicationHelper
+module UsersHelper
+  def gravatar_for user
+        gravatar_id = Digest::MD5::hexdigest user.email.downcase
+        gravatar_url = "https://secure.gravatar.com/avatar/#{gravatar_id}"
+        image_tag gravatar_url, alt: user.name, class: "gravatar"
+  end
```
Comment (saobv-2223) [inline app/helpers/application_helper.rb]: indenting 2 spaces

File: app/helpers/users_helper.rb [hunk 17]
```diff
@@ -0,0 +1,2 @@
+module UsersHelper
+end
```
Comment (saobv-2223) [inline app/helpers/users_helper.rb]: remove file that never used

File: config/routes.rb [hunk 18]
```diff
@@ -1,7 +1,11 @@
 Rails.application.routes.draw do
-  scope "(:locale)", locale: /en/ do
-    root "static_pages#home"
-    get "static_pages/help"
-    get "static_pages/contact"
-  end
+  get "users/show"
+  get "users/new"
```
Comment (saobv-2223) [inline config/routes.rb:4]: move below root

**Comment by NurAnisAqilah** [comment]
ready

**Comment by vanvtt-0952** [comment]
merged

**Pull request merged**